### ISSUE TITLE
write to flash on GW5A is not yet supported

### ISSUE THREAD

**Issue by chopin1998** [trabucayre/openFPGALoader#389]
i got a tang mega 138k pro which use Gowin GW5AST 138 as main fpga chip,

but now openFPGALoader not supported to write its external flash, will support soon?

Thank you!

**Comment by trabucayre** [comment]
I will receive a mega in few days (or a week), will be able to work on adding external flash support.

**Comment by chopin1998** [comment]
got your board ? ^_^

**Comment by trabucayre** [comment]
Yes! I have received my board and started to see how to access flash for this family. It's not really complex (a bit tricky) but I have first to review a PR for gowin devices.

**Comment by chopin1998** [comment]
Good news! i’m a newbie on this, thanks for your help!

**Comment by chopin1998** [comment]
i noticed that openFPGAloader support flash GW5A now! so happy!

but seems write to SRAM function bad now.... when Load a new fs file to SRAM, nothing happen, and READY led on board keeping ON, which should be off

**Comment by trabucayre** [comment]
Currently SPI write is only supported for GW5A, not GW5AST (I don't know why but I have strange behaviour).
It's weird: I have tried yesterday with tangMega and not seen issue when trying to load SRAM, maybe device is in wrong state...
I have to investigate.

**Comment by trabucayre** [comment]
I have also to update status display: currently bit/function are displayed according to non GW5A(ST). For this family only GW5A-25 has the correct register decoding.

**Comment by chopin1998** [comment]
> Currently SPI write is only supported for GW5A, not GW5AST (I don't know why but I have strange behaviour). It's weird: I have tried yesterday with tangMega and not seen issue when trying to load SRAM, maybe device is in wrong state... I have to investigate.

but i can write flash on my mega138pro..

**Comment by trabucayre** [comment]
Ok. So I have to recheck everything. SPI flash is seen but not all the time, it fails to write and I'm able to load SRAM. it's more or less totally opposite situation.
Could you try with `openFPGALoader --bulk-erase` to see if after your able to load a bitstream (ie device is in wrong state due to an issue when bitstream was written).

**Comment by chopin1998** [comment]
when device under wrong state, other write is not work, even with --bulk-erase, need to repower it.

**Comment by trabucayre** [comment]
I continue to search the reason to this regression. before this [commit](https://github.com/trabucayre/openFPGALoader/commit/f71858f96a75c7c0a96f31a4ed3f167e4b914ef0) there is no issue, after yes.
I have some results but it's not really stable and OK/KO depends on bitstream written to the flash...

**Comment by wtjcn** [comment]
I have the same problem, when load to sram of 138k, nothing happen

**Comment by trabucayre** [comment]
I try, as much as possible, to find why SRAM is not working. But I can't isolate the issue. It's a WIP.

**Comment by trabucayre** [comment]
A strange thing I observe: depending of the bitstream written in flash, loading to SRAM works or not.
If I have one of the @sipeed bitstream: SRAM fails. With LiteX bitstream: SRAM is working.
Maybe something wrong with the security option.

**Comment by trabucayre** [comment]
With my last commit I'm able to load a bitstream all the time (with a flash containing a bitstream where security bit is enable or not).
Could you confirm/infirm my tests ?
Thanks.

**Comment by chopin1998** [comment]
c00l!! 

now SRAM write ~100% work, thank thee very much!

flash write need twice, 1st time failed(need to wait), then 2rd time work fine

**Comment by trabucayre** [comment]
I have observed this issue too. I don't know why but `bulk-erase` works all the time.
Maybe the same trick must be applied on Flash side.

**Comment by wtjcn** [comment]
for me, it not work.
when write sram, it show "Erase SRAM double eraseSRAM", progress to 100% and "DONE", but the board's LEDs light continue in old ways. Both litex and official led.fs .

**Comment by trabucayre** [comment]
litex and led.fs uses leds in a similar way. Could you try with litex and [ws2812](https://github.com/sipeed/TangMega-138KPro-example/tree/main/ws2812). I use this demo to have real distinct behavior when I have to switch.
But If after the load you have something (ie no errors) it's a good news.

**Comment by wtjcn** [comment]
1) I found I used linux's FT2232 driver, because it can give /dev/ttyUSB device, although it does not realy work on mega138k
2) I run Gowin_USB_Cable_Installer.sh and then reboot, /dev/ttyUSB disappear, but Gowin's Programmer works well, both for SRAM and FLASH. In this condition, I use openFPGAloader,  sram erase dose not work, and sram programming "DONE" but not work. Another small problem, openFPGAloader default use 6MHz jtag freq,  Gowin's doc said it should not more than 2.5MHz, I change it to 2.5MHz but no change.

**Comment by allrighteveryday** [comment]
In version 0.12.1，Ubuntu , I also meet the problem: nothing happened after burning and READY led on board keeping ON, which should be off.
